src/backups.py

```diff
@@ -0,0 +1,499 @@
+            self.charm._mysql.set_instance_option("tag:_hidden", "true")
+        except MySQLSetInstanceOptionError:
+            self.charm.unit_peer_data["cluster-state"] = "active"
+            return False, "Error setting instance option tag:_hidden"
+
+        try:
+            logger.info("Setting unit as offline before performing backup")
+            self.charm._mysql.set_instance_offline_mode(True)
+        except MySQLSetInstanceOfflineModeError:
+            self.charm.unit_peer_data["cluster-state"] = "active"
+            self.charm._mysql.set_instance_option("tag:_hidden", "false")
+            return False, "Error setting instance as offline before performing backup"
+
+        return True, None
+
+    def _backup(self, backup_path: str, s3_parameters: Dict) -> Tuple[bool, str]:
+        """Runs the backup operations."""
+        try:
+            logger.info("Running the xtrabackup commands")
+            stdout, stderr = self.charm._mysql.execute_backup_commands(
+                s3_parameters["bucket"],
+                backup_path,
+                s3_parameters["access-key"],
+                s3_parameters["secret-key"],
+                s3_parameters["endpoint"],
+            )
+        except MySQLExecuteBackupCommandsError as e:
+            self._upload_logs_to_s3(
+                "",
+                e.message,
+                f"{backup_path}.backup.log",
+                s3_parameters,
+            )
+            return False, "Error backing up the database"
+
+        if not self._upload_logs_to_s3(
+            stdout,
+            stderr,
+            f"{backup_path}.backup.log",
+            s3_parameters,
+        ):
+            return False, "Error uploading logs to S3"
+
+        return True, None
+
+    def _post_backup(self) -> Tuple[bool, str]:
+        """Runs operations required after performing a backup."""
+        logger.info("Setting cluster state as 'active'")
+        self.charm.unit_peer_data["cluster-state"] = "active"
+
+        try:
+            logger.info("Unsetting unit as offline after performing backup")
+            self.charm._mysql.set_instance_offline_mode(False)
+        except MySQLSetInstanceOfflineModeError:
+            return False, "Error unsetting instance as offline before performing backup"
+
+        try:
+            logger.info("Setting unit option tag:_hidden as false")
+            self.charm._mysql.set_instance_option("tag:_hidden", "false")
+        except MySQLSetInstanceOptionError:
+            return False, "Error setting instance option tag:_hidden"
+
+        return True, None
+
+    # ------------------ Perform Restore ------------------
+
+    def _on_restore_backup(self, event: ActionEvent) -> None:
+        """Handle the restore backup action event.
+
+        Restore a backup from S3 (parameters for which can retrieved from the
+        relation with S3 integrator).
+        """
+        backup_id = event.params.get("backup-id")
```

Comment:
might be nice to check for s3-relation

Reply:
addressed in f03e92b4e45ec051635ce73a25ba64629e0eaa03